### ISSUE TITLE
LG-333: Liberty does not limit the age of ffdc log files

### ISSUE THREAD

**Issue by NottyCode** [OpenLiberty/open-liberty#19418]
## Description of the high level feature, including any external spec links:  
The number of files in ffdc directory can grow to very large. Perhaps there should be an option to purge FFDC files when they reach a certain age.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## When complete & mandatory, add links to the UFO (Upcoming Feature Overview) document, FTS (Feature Test Summary), blogs post issues(s), and [Aha](https://cloud-platform.ideas.aha.io/) (externally raised RFEs):
- UFO: https://ibm.ent.box.com/file/987944122537
- FTS: https://github.com/OpenLiberty/open-liberty/issues/23454
- Beta Blog Post (if applicable): https://github.com/OpenLiberty/open-liberty/issues/23235
- Blog Post:  https://github.com/OpenLiberty/open-liberty/issues/23614
- Aha: https://cloud-platform.ideas.ibm.com/ideas/TWAS-I-145
   - Also, add a comment into the Aha idea linking to this issue.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)

Instructions:
- Do the actions below and mark them complete in the checklist when they are done.
- Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.

## **Design**
Before Development Starts or 8 weeks before Onboarding
- [x] POC Design / UFO Review Scheduled (David Chang) or N/A.
- [x] POC Design / UFO Reviewed (Feature Owner) or N/A.
- [x] Complete any follow-ons from the POC Review.
- [x] Design / UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
- [x] No Design / No UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
  * Note: Design is required if any of the following apply: SVT or Performance testing, new UI, Servicibility considerations, significant documentation changes
- [x] ID Requirements identified ([Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty)). (Epic owner / Feature owner with ID focal point)
  * Note: If only trivial documentation changes are required, you may reach out to the ID Feature Focal to request a `ID Required - Trivial` label.  Unlike features with regular ID requirement, those with `ID Required - Trivial` label do not have a hard requirement for a Design/UFO.
- [x] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [x] Create a child task of this epic entitled "Feature Test Summary" via [this template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=). Add the link in above.

##
##

## **Development**
When active development has begun 
- [x] Add the "In Progress" label to this issue.

## **Beta**
If your feature, or portions of it, are going to be included in a beta  
Before Onboarding the beta
- [x] Beta Fence the functionality (`kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`)  

1 week before beta GA
- [x] Create, populate, and link to the [Beta blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_beta.md&title=BETA+BLOG+-+title_of_your_update)

## **Legal**
3 weeks before Onboarding
- [x] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.

## **Translation**
3 weeks before Onboarding
- [x] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)

## **Feature Complete**
2 weeks before Onboarding
- [x] Implementation complete. (Epic owner / Feature owner)
- [x] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [x] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)

## **Focal Point Approvals**
2 to 1 week before Onboarding
#### You **MUST** have the Design Approved or No Design Approved label before requesting focal point approvals.

All features (both "Design Approved" and "No Design Approved")
- [x] **FAT** - ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers)). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [x] **Demo** - ([Tom Evans](https://github.com/tevans78) or [Chuck Bridgham](https://github.com/cbridgha)). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
- [x] **Globalization** ([Sam Wong](https://github.com/samwatibm) - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.

"Design Approved" features
- [x] **Accessibility** - ([Steven Zvonek](https://github.com/steven1046)). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [x] **ID** - ([Karen Deen](https://github.com/chirp1)). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
  * Note: If only trivial documentation changes are required, you may reach out to the ID Feature Focal to request a `ID Required - Trivial` label.  Unlike features with regular ID requirement, those with `ID Required - Trivial` label do not have a hard requirement for a Design/UFO.
- [x] **Performance** - ([Jared Anderson](https://github.com/jhanders34)). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [x] **Serviceability** - ([Don Bourne](https://github.com/donbourne)). Serviceability has been addressed.
- [x] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [x] **SVT** - ([Brian Hanczaryk](https://github.com/hanczaryk) - APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.

## **Ready for GA**
1 week before Onboarding
- [x] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [x] Github Epic and Epic's issues are closed / complete. All PRs are committed to the release branch. (Epic owner / Feature owner / Backlog Subtribe PM)

1 week before GA
- [x] Create, populate, and link to the [Blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_ga_release.md&title=GA+BLOG+-+title_of_your_update)

## **Other deliverbles**
- [ ] **OL Guides** - ([Yee-Kang Chang](https://github.com/yeekangc)). Assessment for OL Guides is complete or N/A.
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
- [ ] **Blog** - ([Laura Cowen](https://github.com/lauracowen)) Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by realModusOperandi** [comment]
UFO Notes
- Slide 12: maxFfdcAge should use a Duration for metatype
- Slide 29: Migration tools should bring across FFDC deletion age from tWAS to Liberty
- End: Consider upgrading existing message when FFDCs are deleted (possibly including how many) from debug to info, without listing every file deleted (that can be trace)
- End: Consider allowing the time for the deletion to be configured by the user

**Comment by NottyCode** [comment]
I think the intent is if maxFfdcAge is set we will delete only based on the age, and not based on writing 500 files, but I don't see it explicitly stated in the UFO. I would like it to be there.

There is a link in the UFO to an RFE, but those RFEs are going away. It should link to https://cloud-platform.ideas.ibm.com/ideas/TWAS-I-145 instead. I've added this into the description of this item which it should have been in.

There was a discussion about writing a message when purge processing runs, but this didn't get reflected in a UFO update.

There was a discussion about what happens if the server isn't running at midnight, but I don't see anything reflected in the UFO for that discussion. It is possible in the discussion we agreed that we wouldn't do anything with this, but I'd like to see a comment in the issue that indicates things from the socialization that aren't being actioned and why.

I like the short design though.

**Comment by tonyreigns** [comment]
Hi Alasdair, I don't think we discussed not deleting files that reached over the 500 cap if maxFfdcAge was set. I was under the impression that that the 500 file cap would be left alone aside from just changing the log level from debug to info which I was planning on doing. 

We chose not to add an additional variable to set the time of the purge to prevent the clutter of extra variables that may not be used, but it could easily be added if there was demand for it. I'll make sure to reflect all these changes/updates in the UFO. 

Thanks.

**Comment by NottyCode** [comment]
> Hi Alasdair, I don't think we discussed not deleting files that reached over the 500 cap if maxFfdcAge was set. I was under the impression that that the 500 file cap would be left alone aside from just changing the log level from debug to info which I was planning on doing.

I agree it wasn't discussed on the call, but I want to have it clearly stated in the design how the two modes work together. I'm guessing it is a by date or by number but that isn't explicit so it could be over that age if there are more than 500. I'd like that aspect of the design spelled out.

What about the other two comments I made?

**Comment by tonyreigns** [comment]
I understand, I've updated the UFO with everything you've requested. Aside from the Aha link, the changes are reflected on page 13 of the UFO.

There'll still be a hard cap at 500 and FFDC files could still be deleted before the maxFfdcAge is reached if the number of files reaches 500. 

I've also mentioned the message during the purge, the Aha link, what happens if the server isn't running at midnight and why we chose to not action it.

Thanks.

**Comment by tonyreigns** [comment]
Documentation issue opened: https://github.com/OpenLiberty/docs/issues/6058

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

5.  Does this feature add any new metrics or emit any new JSON events?  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?

**Comment by tonyreigns** [comment]
1.  UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?
**Yes, the UFO identifies problems customer can see. An error message will be logged when an invalid duration values is entered which allows customers to diagnose and solve this problem on their own.**

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
**Invalid maxFfdcAge values were entered which emits FFDC errors.**
b) Who did you demo to?
**Morgan Li**
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?
**Yes**

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
**It was agreed no SVT is required for this feature. FAT coverage is sufficient.**
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  
**Yes**

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.
**WAS L3: Liberty Log Analytics**

5.  Does this feature add any new metrics or emit any new JSON events?  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?
**No**

**Comment by tonyreigns** [comment]
@OpenLiberty/performance-approvers  We don't expect a performance hit in this feature, I'd appreciate an approval, thanks.

**Comment by tonyreigns** [comment]
@OpenLiberty/id-approvers The documentation issue was opened opened last week: https://github.com/OpenLiberty/docs/issues/6058 

Is anything else required? Thanks.

**Comment by chirp1** [comment]
ID has the information that they need in the doc issue at https://github.com/OpenLiberty/docs/issues/6058 . Approving the epic.

**Comment by tonyreigns** [comment]
All approvals have been received and the beta fence has been removed.

**Comment by tonyreigns** [comment]
Blog post: https://github.com/OpenLiberty/open-liberty/issues/23614
